app/services/course/assessment/submission/auto_grading_service.rb

```diff
@@ -65,4 +66,22 @@ def aggregate_failures(jobs)
     error_messages = failed_jobs.map { |job| job.error['message'] }
     raise SubJobError, error_messages.to_sentence
   end
+
+  def publish_grade(submission)
+    submission.answers.select(&:submitted?).each(&:publish!)
+    submission.points_awarded = calculate_exp(submission)
+    submission.publish!
+  end
+
+  # Calculating scheme:
+  #   Submit in time: ( base_exp + bonus_exp ) * actual_grade / max_grade
+  #   Late submission: base_exp * actual_grade / max_grade
+  def calculate_exp(submission)
+    assessment = submission.assessment
+    total_exp = assessment.base_exp
+    if !assessment.bonus_end_at || submission.submitted_at <= assessment.bonus_end_at
+      total_exp += assessment.time_bonus_exp
+    end
+    (submission.grade.to_f / assessment.maximum_grade * total_exp).to_i
```

Comment:
We can currently set questions to have a maximum grade of 0. Should we return full exp if assessment max grade is zero?